dask/array/random.py

```diff
@@ -48,31 +49,101 @@ def _wrap(self, func, *args, **kwargs):
 
         extra_chunks should be a chunks tuple to append to the end of chunks
         """
-        size = kwargs.pop('size')
+        size = kwargs.pop('size', None)
         chunks = kwargs.pop('chunks')
         extra_chunks = kwargs.pop('extra_chunks', ())
 
-        if not isinstance(size, (tuple, list)):
+        if size is not None and not isinstance(size, (tuple, list)):
             size = (size,)
 
+        args_shapes = {ar.shape for ar in args
+                       if isinstance(ar, (Array, np.ndarray))}
+        args_shapes.union({ar.shape for ar in kwargs.values()
+                           if isinstance(ar, (Array, np.ndarray))})
+
+        shapes = list(args_shapes)
+        if size is not None:
+            shapes += [size]
+        # broadcast to the final size(shape)
+        size = broadcast_shapes(*shapes)
         chunks = normalize_chunks(chunks, size)
+        slices = slices_from_chunks(chunks)
+
+        def _broadcast_any(ar, shape, chunks):
+            if isinstance(ar, Array):
+                return broadcast_to(ar, shape).rechunk(chunks)
+            if isinstance(ar, np.ndarray):
+                return np.broadcast_to(ar, shape)
+
+        # Broadcast all arguments, get tiny versions as well
+        # Start adding the relevant bits to the graph
+        dsk = {}
+        lookup = {}
+        small_args = []
+        for i, ar in enumerate(args):
+            if isinstance(ar, (np.ndarray, Array)):
+                res = _broadcast_any(ar, size, chunks)
+                if isinstance(res, Array):
+                    dsk.update(res.dask)
+                    lookup[i] = res.name
+                elif isinstance(res, np.ndarray):
+                    name = 'array-{}'.format(tokenize(res))
```

Comment:
If you really wanted to ensure same names we could restride everything before we tokenize.  This could happen here or (with a bit of profiling) within the `normalize_token(np.ndarray)` function itself.